### PR TITLE
systemd: remove hardcoded ntp servers

### PR DESCRIPTION
**Pull request by emhamm**
Ntp servers defaults for `systemd-timesyncd` are set in `options.networking.timeServers` which again defaults to  `0.nixos.pool.ntp.org 1.nixos.pool.ntp.org 2.nixos.pool.ntp.org 3.nixos.pool.ntp.org`

At the same time these same servers are getting hard compiled into systemd-timesyncd when systemd gets build and those can not be deleted nor changed with config options, which is not desirable. 

Especially in networks where those time servers can not be reached and none is provided this creates a lot of churn in the system logs.

This change removes the non-configurable ntp server addresses compiled into systemd-timesyncd.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by arianvp** [comment]
This doesn't do what you think it does.

If we don't specify `ntp-servers` system defaults to:


```
option('ntp-servers', type : 'string',
       description : 'space-separated list of default NTP servers',
       value : 'time1.google.com time2.google.com time3.google.com time4.google.com')
       ```
```

Note that the compile-time option is only used as a fallback in case there is no explicit ntp configuration. You can still configure alternatives.

So

> when systemd gets build and those can not be deleted nor changed with config options, which is not desirable.

is not true. If you specify a config; the fallback isn't used.


See https://github.com/systemd/systemd/blob/main/docs/DISTRO_PORTING.md#ntp-pool

> By default, systemd-timesyncd uses the Google Public NTP servers time[1-4].google.com, if no other NTP configuration is available. They serve time that uses a [leap second smear](https://developers.google.com/time/smear) and can be up to .5s off from servers that use stepped leap seconds.
> If you prefer to use leap second steps, please register your own vendor pool at ntp.org and make it the built-in default by passing -Dntp-servers= to meson. Registering vendor pools is [free](http://www.pool.ntp.org/en/vendors.html).
> Use -Dntp-servers= to direct systemd-timesyncd to different fallback NTP servers.

**Comment by emhamm** [comment]
@arianvp thanks for your reply and the link.

You are right, if the fallback ntp servers are not provided then a google timeserver fallback is in place and this is not what I intended. 
The added switch makes this configurable at build time, if set an empty string is provided to meson and no fallback servers are set, no noise in the logs about failed or timed out synchronizations.
This is now the desired behaviour.

**Comment by flokli** [comment]
I still don't get what your goal is, and don't think adding more config knobs to the systemd derivation is the right way to go.

Why patching ntp servers out of systemd? If you don't want timesyncd to do NTP, simply disable it via NixOS configuration:
`services.timesyncd.enable`. If you want to set you own custom NTP servers, use `services.timesyncd.servers`.

**Comment by emhamm** [comment]
@flokli I shall clarify this a bit more:

I already provide `services.timesyncd.servers = [ ]` so no ntp servers shall be set. Nevertheless there are still those fallback ntp servers.
There are networks in which those preconfigured fallback ntps are not reachable and no others are reachable and/or configured. In this case there is quite a lot of noise in the system log where systemd-timesyncd complains about failed or timed out ntp requests.

**Comment by flokli** [comment]
> @flokli I shall clarify this a bit more:
> 
> I already provide `services.timesyncd.servers = [ ]` so no ntp servers shall be set. Nevertheless there are still those fallback ntp servers. There are networks in which those preconfigured fallback ntps are not reachable and no others are reachable and/or configured. In this case there is quite a lot of noise in the system log where systemd-timesyncd complains about failed or timed out ntp requests.

Well, in this case, if you have no time servers available, just disable systemd-timesyncd with `services.timesyncd.enable`. What's the point in having it running if you have no time servers at all?

**Comment by emhamm** [comment]
Sounds silly, i get it. 

I can not control in which environment those systems will be deployed and there is a way for users to set NTP servers on a running system. They are not enabled to change the nixos config nor can I if there is simply no way to connect to the systems. So I may not simply disable timesyncd and go on. If they choose not to configure any NTP servers, the fallback servers are used and then their systemlog gets cluttered with those error messages. To disable those fallback servers solves this.

**Comment by flokli** [comment]
Then please ship this change to `mesonFlagsArray` in a local override to `systemd.package` on these systems, rather than setting `withFallbackNtpServers = false` there.

**Comment by emhamm** [comment]
in plain english: 

there are defaults set in `options.networking.timeServers`, which may be overriden, but if deleted they reapear. 
This behaviour is not documented, it is just plain surprising and I call this a bug.

I tend to say the standard systemd-timesync config with their defaults does not apply here, because defaults are already set in nix options. What should be set is an empty meson array, because default ntp servers are configured in nixos config and not at compile time where it can not be changed with the options supposed to change this setting.

**Comment by flokli** [comment]
The compiled-in defaults are how systemd is compiled for this distribution. If you don't like the defaults, you're free to override the systemd package locally, but that's not something nixpkgs should have custom logic for.

If you don't want to have timesyncd running, disable it, rather than configuring an empty list of servers. https://github.com/NixOS/nixpkgs/pull/296924 further points out this behaviour in the documentation for this option.
